packages/contracts-bedrock/src/L1/opcm/OPContractsManagerV2.sol

```diff
@@ -0,0 +1,1194 @@
+///         release.
+/// @dev When adding a new dispute game type, if your dispute game requires configuration that
+///      differs from configuration used by other dispute game types, you will need to add a new
+///      configuration struct and then add parsing logic for that struct in the _makeGameArgs
+///      function. You will also need to return the correct game implementation in _getGameImpl.
+/// @dev When adding a net-new input, simply add the input to FullConfig and add the corresponding
+///      logic for loading that input in _loadFullConfig. NOTE that when adding a completely new
+///      input, users upgrading an existing chain will need to supply that input in the form of an
+///      override as part of the UpgradeInput struct.
+/// @dev If you were going to build a V3 of OPCM, you probably want to make this look a lot more
+///      like Terraform. The V2 design is trending in the direction of being Terraform-like, but it
+///      doesn't quite get there yet in an attempt to be a more incremental improvement over the V1
+///      design. Look at _execute, squint, and imagine that it can output an upgrade plan rather
+///      than actually executing the upgrade, and then you'll see how it can be improved.
+contract OPContractsManagerV2 is ISemver {
+    /// @notice Configuration struct for the FaultDisputeGame.
+    struct FaultDisputeGameConfig {
+        Claim absolutePrestate;
+    }
+
+    /// @notice Configuration struct for the PermissionedDisputeGame.
+    struct PermissionedDisputeGameConfig {
+        Claim absolutePrestate;
+        address proposer;
+        address challenger;
+    }
+
+    /// @notice Generic dispute game configuration data.
+    struct DisputeGameConfig {
+        bool enabled;
+        uint256 initBond;
+        GameType gameType;
+        bytes gameArgs;
+    }
+
+    /// @notice Contracts that represent the Superchain system.
+    struct SuperchainContracts {
+        ISuperchainConfig superchainConfig;
+    }
+
+    /// @notice Contracts that represent the full chain system.
+    struct ChainContracts {
+        ISystemConfig systemConfig;
+        IProxyAdmin proxyAdmin;
+        IAddressManager addressManager;
+        IL1CrossDomainMessenger l1CrossDomainMessenger;
+        IL1ERC721Bridge l1ERC721Bridge;
+        IL1StandardBridge l1StandardBridge;
+        IOptimismPortal optimismPortal;
+        IETHLockbox ethLockbox;
+        IOptimismMintableERC20Factory optimismMintableERC20Factory;
+        IDisputeGameFactory disputeGameFactory;
+        IAnchorStateRegistry anchorStateRegistry;
+        IDelayedWETH delayedWETH;
+    }
+
+    /// @notice Struct that represents an additional instruction for an upgrade. Each upgrade has
+    ///         its own set of extra upgrade instructions that may or may not be required. We use
+    ///         this struct to keep the upgrade interface the same each time.
+    struct ExtraInstruction {
```

Comment:
I want to be sure that I understand this. 

So we have the requirement that: "Teams should be able to work with OPCM at feature development time without worrying about coming back to clean code up again in the future."

So, if we are adding a new initializer value in upgrade 69, then that new value will need to be provided as an `ExtraInstruction` on `UpgradeInput` when performing the upgrade. However for upgrade 70, you don't need to come back and edit the code that uses the `ExtraInstruction`, because you used `loadBytes()` and `loadFullConfig()` so that the default is to read the value from the contract, unless an override is provided.

Reply:
That's the idea, yes. You do unfortunately still need to come back in upgrade 70 to clear out the `_assertValidUpgradeInstructions` field.